executor/executor.go

```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
 	if err := e.baseExecutor.Open(ctx); err != nil {
 		return errors.Trace(err)
 	}
+	for _, src := range e.srcs {
+		if err := src.Open(ctx); err != nil {
+			return errors.Trace(err)
+		}
+	}
 	e.done = false
 	return nil
 }
 
-// Next implements the Executor Next interface.
-func (e *CheckTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+// Close implements the Executor Close interface.
+func (e *CheckTableExec) Close() error {
 	if e.done {
 		return nil
 	}
-	defer func() { e.done = true }()
-	for _, t := range e.tables {
-		dbName := t.DBInfo.Name
-		tb, err := e.is.TableByName(dbName, t.Name)
+
+	close(e.exitCh)
+	var err error
+	for _, src := range e.srcs {
+		err1 := src.Close()
+		if err == nil && err1 != nil {
+			err = err1
+		}
+	}
+	return errors.Trace(err)
+}
+
+func (e *CheckTableExec) checkIndexHandle(ctx context.Context, num int, src *IndexLookUpExecutor) error {
+	cols := src.schema.Columns
+	retFieldTypes := make([]*types.FieldType, len(cols))
+	for i := range cols {
+		retFieldTypes[i] = cols[i].RetType
+	}
+	chk := chunk.New(retFieldTypes, e.initCap, e.maxChunkSize)
+	iter := chunk.NewIterator4Chunk(chk)
+
+	var err error
+	handles := make(map[int64]struct{}, 1024)
+	for {
+		err = src.Next(ctx, chk)
 		if err != nil {
-			return errors.Trace(err)
+			break
 		}
-		if tb.Meta().GetPartitionInfo() != nil {
-			err = e.doCheckPartitionedTable(tb.(table.PartitionedTable))
-		} else {
-			err = e.doCheckTable(tb)
+		if chk.NumRows() == 0 {
+			break
 		}
-		if err != nil {
-			log.Warnf("%v error:%v", t.Name, errors.ErrorStack(err))
-			if admin.ErrDataInConsistent.Equal(err) {
-				return ErrAdminCheckTable.GenWithStack("%v err:%v", t.Name, err)
+
+		for chunkRow := iter.Begin(); chunkRow != iter.End(); chunkRow = iter.Next() {
+			datum := chunkRow.GetDatum(0, retFieldTypes[0])
+			h := datum.GetInt64()
+			if _, ok := handles[h]; ok {
+				err = errors.Errorf("admin check table %s, index %s handle %d more than one", e.tblInfo.Name, e.indices[num].Meta().Name, h)
+				e.retCh <- errors.Trace(err)
+				return errors.Trace(err)
 			}
+			handles[h] = struct{}{}
+		}
 
-			return errors.Errorf("%v err:%v", t.Name, err)
+		select {
+		case <-e.exitCh:
+			return nil
+		default:
 		}
 	}
-	return nil
+	e.retCh <- errors.Trace(err)
+	return errors.Trace(err)
 }
 
-func (e *CheckTableExec) doCheckPartitionedTable(tbl table.PartitionedTable) error {
-	info := tbl.Meta().GetPartitionInfo()
-	for _, def := range info.Definitions {
-		pid := def.ID
-		partition := tbl.GetPartition(pid)
-		if err := e.doCheckTable(partition); err != nil {
+// Next implements the Executor Next interface.
+func (e *CheckTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
```

Comment:
Seems we only check the index handle value except for the number of rows on the table is greater than the index.
I think it is better to check the index values.

Reply:
I have asked your opinions before, you feel that you can accept not check index value. If we really need it, we can handle it in the next PR.

Reply:
I used to think your question is about the error message with the handle only or with the values. I think it is my fault to misunderstand your question.